src/XrdHttp/XrdHttpReq.cc

```diff
@@ -1798,7 +1798,16 @@ XrdHttpReq::PostProcessChecksum(std::string &digest_header) {
     if (convert_to_base64) {free(digest_value);}
     return 0;
   } else {
-    prot->SendSimpleResp(500, NULL, NULL, "Underlying filesystem failed to calculate checksum.", 0, false);
+    if(xrderrcode != kXR_NotFound) {
+      prot->SendSimpleResp(500, NULL, NULL, "Underlying filesystem failed to calculate checksum.", 0, false);
+    } else {
+      // This is a result of the "Ugly" hack put in place to prevent a failing stat() on a
+      // non-manager server to fail a transfer.
+      // In the case a user issues a GET with a 'Want-Digest' header on a non-existing file, the non-failing
+      // stat() will result in a failed checksum query giving a 500 error back to the user.
+      // Here we just return a 404 instead. httpStatusCode and httpStatusText contain the right reason why the request failed.
+      prot->SendSimpleResp(httpStatusCode, NULL, NULL, httpStatusText.c_str(), httpStatusText.length(), false);
+    }
```

Comment:
```suggestion
    // This is a result of the "Ugly" hack put in place to prevent a failing stat() on a
    // non-manager server to fail a transfer.
    // In the case a user issues a GET with a 'Want-Digest' header on a non-existing file, the non-failing
    // stat() will result in a failed checksum query giving a 500 error back to the user.
    // Here we just return a 404 instead. httpStatusCode and httpStatusText contain the right reason why the request failed.
    if(xrderrcode != kXR_NotFound) {
      httpStatusCode = 500;
      httpStatusText = "Underlying filesystem failed to calculate checksum.";
    }
        
    prot->SendSimpleResp(httpStatusCode, NULL, NULL, httpStatusText.c_str(), httpStatusText.length(), false);
    }
```